diff --git a/kernel/Kconfig.hz b/kernel/Kconfig.hz
index 38ef6d06888e..0f78364efd4f 100644
--- a/kernel/Kconfig.hz
+++ b/kernel/Kconfig.hz
@@ -40,6 +40,27 @@ choice
   on SMP and NUMA systems and exactly dividing by both PAL and
   NTSC frame rates for video and multimedia work.
 
+	config HZ_500
+		bool "500 HZ"
+	help
+	 500 Hz is a balanced timer frequency. Provides fast interactivity
+	 on desktops with good smoothness without increasing CPU power
+	 consumption and sacrificing the battery life on laptops.
+
+	config HZ_625
+		bool "625 HZ"
+	help
+	 625 Hz is a balanced timer frequency. Provides fast interactivity
+	 on desktops with good smoothness without increasing CPU power
+	 consumption and sacrificing the battery life on laptops.
+
+	config HZ_750
+		bool "750 HZ"
+	help
+	 750 Hz is a balanced timer frequency. Provides fast interactivity
+	 on desktops with good smoothness without increasing CPU power
+	 consumption and sacrificing the battery life on laptops.
+
  config HZ_1000
    bool "1000 HZ"
  help
@@ -53,6 +74,9 @@ config HZ
  default 100 if HZ_100
  default 250 if HZ_250
  default 300 if HZ_300
+	default 500 if HZ_500
+	default 625 if HZ_625
+	default 750 if HZ_750
  default 1000 if HZ_1000
 
 config SCHED_HRTICK
